### ISSUE TITLE
Add self & target conditions to damage parts

### ISSUE THREAD

**Issue by arbron** [foundryvtt/dnd5e#4477]
The new data structure for damage gives the system an opportunity to define more specific data for each damage part. An extremely useful feature would be conditional damage parts. These conditions would come in two flavors: self conditions based on the actor rolling the damage, and target conditions based on the actor taking the damage.

### Self Conditions
These conditions apply to data on the actor rolling the damage and are evaluated before the damage roll is constructed. These conditions can include things like the actor's current HP (only do this damage if below 25% health), size (extra damage if big), status effects (reduced damage if swarm is bloodied), etc. These can also use the item state, so a weapon like a Flametongue can have some sort of state indicating whether it is ignited or not, and only roll the extra fire damage if it is.

### Target Conditions
These conditions are evaluated during the `calculateDamage` step on a per-target basis. This means that these damage parts are always rolled as part of the damage. The same actor-specific conditions that are available for self conditions are usable here, allowing for limiting certain damage by creature type (extra damage vs. undead), hit points (one die size against health creatures, another against damage creatures), etc.

### Data Structure
The data structure defining these conditions would be a fairly simple array of objects:

```javascript
{
  target: "target",  // Can be `item`, `self`, or `target` to indicate what data is checked
  type: "creature-type",
  filter: { k: "system.details.type.value", v: "undead" }
}
```

This is a filter indicating the damage only applies if the target is undead. The filter object is a `FilterDescription` (this data structure is already in use by the system for the Compendium Browser) which allows for nested filter conditions allowing for very complex queries that can then be directly evaluated against target. For example the above filter could be wrapped in a `NOT` filter to invert it:

```javascript
{ o: "NOT", v: { k: "system.details.type.value", v: "undead" } }
```

It could also be changed to an `in` filter to indicate more than one creature type:

```javascript
{ k: "system.details.type.value", o: "in", v: ["field", "undead"] }
```

### Module Compatibility & UI
The `type` value in the above filter is used to control how the UI for the condition is displayed. In this case the system will provide a number of specific UIs to set conditions for certain actor properties so users don't have to dig down into specific attribute keys. Module should be able to provide their own condition types that display their own UIs.

Since everything ultimately is stored as a single `FilterDescription` object behind the scenes, this protects against problems if a module is disabled since we can still evaluate these conditions regardless of if the module-provided UI is available.

**Comment by krbz999** [comment]
Such data should not be added to only items. Ideally it would be on items, on effects, and on actors as well (just nuke the `system.bonuses` property, it's outdated).

There are cases where an actor would want a more global, additional damage part. For example a cleric adding their wisdom modifier to all cantrips. Users shouldn't be required to modify every cantrip they add to their sheet.

I would make the same argument for dice modifiers, not only additional damage parts. For example the Elemental Adept feat, which modifies all dice if of a certain damage type.

There's a need for a better system than *just* conditional damage parts.

**Comment by arbron** [comment]
> There's a need for a better system than _just_ conditional damage parts.

True, but that is beyond the scope of this issue. Those other modifiers would exist on the actor (potentially applied by items using AEs, but living on the actor nonetheless), while this a much narrowly scoped issue specific for activity damage parts.

**Comment by kaelad02** [comment]
If active effects could be a filter, not just status effects, that could handle spells like Hunter's Mark and Hex and their bonus damage.

This might be out of scope, but what about optional damage like Sneak Attack that's manually applied? It wouldn't be conditional based on the source or target, but more like a checkbox to apply the damage.
